autopush/jwt.py

```diff
@@ -0,0 +1,102 @@
+import base64
+import binascii
+import json
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import ec
+from cryptography.hazmat.primitives import hashes
+
+from pyasn1.type import univ, namedtype
+from pyasn1.codec.der.encoder import encode
+
+
+"""Why hand roll?
+Most python JWT libraries either use a python elliptic curve library directly,
+or call one that does, or is abandoned, or a dozen other reasons.
+
+After spending half a day looking for reasonable replacements, I decided to
+just write the functions we need directly.
+
+THIS IS NOT A FULL JWT REPLACEMENT.
+
+"""
+
+
+def repad(string):
+    # type: (str) -> str
+    """Adds padding to strings for base64 decoding"""
+    if len(string) % 4:
+        string += '===='[len(string) % 4:]
+    return string
+
+
+class VerifyJWT:
+
+    def __init__(self):  # pragma: no cover
+        pass
+
+    @classmethod
+    def raw_sig_to_der(cls, auth):
+        # type: (cls, str) -> str
+        """Fix the JWT auth token.
+
+        The `ecdsa` library signs using a raw, 32octet pair of values (r,s).
+        Cryptography, which uses OpenSSL, uses a DER sequence of (s, r).
+        This function converts the raw ecdsa to DER.
+
+        """
+        payload, asig = auth.rsplit(".", 1)
+        sig = base64.urlsafe_b64decode(repad(asig).encode('utf8'))
+        if len(sig) != 64:
+            return auth
+
+        # ecdsa and openssl transpose the "r" and "s" values of the signatures.
+        # for ecdsa signature is (r, s)
+        # for openssl, signature is (s, r)
+        # It's ok, though, because neither label them, even though openssl
+        # uses namedtypes, with the names set to ""
+        #
+        # Oh frabjous day!
+        ds = univ.Sequence(
+            componentType=namedtype.NamedTypes()
+        ).setComponents(
+            univ.Integer(int(binascii.hexlify(sig[:32]), 16)),
+            univ.Integer(int(binascii.hexlify(sig[32:]), 16)),
+        )
+        encoded = encode(ds)
+        new_sig = base64.urlsafe_b64encode(encoded)
+        return "{}.{}".format(payload, new_sig)
```

Comment:
Since the caller immediately calls `split()` on this, is there a reason not to return a tuple?

This saves you the splitting, and removes the need to to base64-encode (and then decode) the signature.